ipa-core/src/cli/clientconf.rs

```diff
@@ -72,162 +72,157 @@ pub fn setup(args: ConfGenArgs) -> Result<(), BoxError> {
-            Value::Table(encode_hpke(mk_public_key)),
-        );
-        peers.push(peer.into());
-    }
+    /// Number of shards per helper
+    #[arg(long)]
+    pub shard_count: u32,
 
-    let client_config = if use_http1 {
-        ClientConfig::use_http1()
-    } else {
-        ClientConfig::default()
-    };
-    let mut network_config = Map::new();
-    network_config.insert(String::from("peers"), peers.into());
-    network_config.insert(
-        String::from("client"),
-        Table::try_from(client_config)?.into(),
-    );
-    let config_str = toml::to_string_pretty(&network_config)?;
+    /// Shard port number to be used for all servers.
+    #[arg(long)]
+    pub shards_port: u16,
 
-    // make sure network config is valid
-    if cfg!(debug_assertions) {
-        assert_network_config(&network_config, &config_str);
-    }
+    /// MPC port number  to be used for all servers.
+    #[arg(long)]
+    pub mpc_port: u16,
 
-    Ok(conf_file.write_all(config_str.as_bytes())?)
+    /// Destination folder for the config file. If not specified, `keys_folder` will be used.
+    #[arg(long)]
+    pub(crate) output_dir: Option<PathBuf>,
+
+    /// Overwrite configuration file if it exists at destination.
+    #[arg(long, default_value_t = false)]
+    pub(crate) overwrite: bool,
 }
 
-/// Creates a section in TOML that describes the HPKE configuration for match key encryption.
-fn encode_hpke(public_key: String) -> Table {
-    let mut hpke_table = Table::new();
-    // TODO: key registry requires a set of public keys with their "identifier". Right now
-    // we encode only one key
-    hpke_table.insert(String::from("public_key"), Value::String(public_key));
+/// Similar to [`setup`] but for a sharded setup.
+///
+/// # Errors
+/// If the files aren't in the expected locations, or if the parameters like shard count don't
+/// match.
+pub fn sharded_setup(args: ShardedConfGenArgs) -> Result<(), BoxError> {
+    let clients_conf = create_sharded_conf_from_files(
+        args.shard_count,
+        args.mpc_port,
+        args.shards_port,
+        args.keys_dir.clone(),
+    );
+    gen_conf_from_args(args.output_dir, args.overwrite, args.keys_dir, clients_conf)
+}
 
-    hpke_table
+/// this helper function creates [`HelperClientConf`] for a sharded configuration. We go ring by
+/// ring as expected by helper binary.
+fn create_sharded_conf_from_files(
+    shard_count: u32,
+    port: u16,
+    shard_port: u16,
+    keys_dir: PathBuf,
+) -> impl Iterator<Item = HelperClientConf> {
+    (0..shard_count).flat_map(move |ix| {
+        let base_dir = keys_dir.clone();
+        HelperIdentity::make_three().into_iter().map(move |id| {
+            let mut shard_dir = base_dir.clone();
+            let id_nr: u8 = id.into();
+            shard_dir.push(format!("helper{id_nr}"));
+            shard_dir.push(format!("shard{ix}"));
+
+            let host_name = find_file_with_extension(&shard_dir, "pem").unwrap();
```

Comment:
can we assume that the full file name is known and simplify this? We wouldn't need file search in this case

Reply:
Discussed offline. We use the filename to pass the host name.